### PR TITLE
Support space-only resource register annotations

### LINKED ISSUE #2059: [DXIL] Automatic resource binding with only spaceN specified for conditional compilation
Hello,

I would like to use the DXC reflection layer to obtain the descriptor binding assignment (by not manually assigning them in the shader), so that conditional compilation can eliminate dead resource bindings.
My use case is as follows:
```hlsl
#ifdef _RESOURCE_REQUIRED
Texture2D MyTexture : register(space0);
#else
cbuffer ColorCB : register(space0) {
  float4 MyColor ;
};
#endif

```
My intention is not to specify the binding (register(b0) or register(t0)), but provide the spaceN number so that at-least on the engine level on the Vulkan side, I would need to bind an unchanging (or depending upon the frequency of change) descriptor set once for a frame. More precisely I am reserving slots (N) depending upon the frequency of change of resources (where the layout is matching), and use the reflection layer to get the individual resource binding slots.

The code example actually compiles fine if I use DXC with SPIRV codegen, and the generated assembly is also valid. However, this does not compile fine with DXC and DXIL code gen.
I have also tried with glslang and it throws an error on the register keyword usage.
So for me the grammar I am using is incorrect, although it works for DXC-SPIRV.

With glsl, I can do something like layout(set = N), and let the binding be specified automatically.
Would it be possible to provide this mechanism using a vulkan annotation? Something like:
`[[vk::binding_set(N)]] Texture2D MyTexture`

Thanks for any help.

### PR DESCRIPTION
**Pull request by MrTrillian**
This change adds support for `: register(spaceN)`, where a space is specified but no register, such that the resource allocator runs on the given space.

For backwards compatibility, `-auto-binding-space` applies to resources without register bindings but not to those with an explicit register assignment like `: register(t0)` - those always go in space 0. To support this, I had to add state to represent  "unspecified space". In the `UnusualAnnotation`, it's a `Optional<uint32_t>`. In the `DxilResource`, it's `UINT_MAX`, like we do for unspecified registers.

SPIR-V provided the front-end part of the implementation, though I changed from using `ResourceType == 'x'` as a way to indicate space-only to making it `ResourceType == 0 && ResourceSpace.hasValue()`. 

Fixes #2059

### CODE REVIEW

**Comment by ehsannas** [comment]
Awesome! Thanks @MrTrillian 👍

**Comment by MrTrillian** [comment]
@ehsannas I might need your help here too, unless you'd really rather I didn't introduce this `llvm::Optional<>`

**Comment by ehsannas** [comment]
@MrTrillian Thanks for this PR! It's great to have this feature supported in both backends and make the two worlds align. I will make the necessary update in the SPIR-V backend.

File: tools/clang/include/clang/AST/HlslTypes.h [hunk 1]
```diff
@@ -215,35 +217,22 @@ struct UnusualAnnotation {
...
+  uint32_t          RegisterOffset = 0;
 
   void setIsValid(bool value) {
     IsValid = value;
   }
 
-  void setAsSpaceOnly() { RegisterType = 'x'; }
-  bool isSpaceOnly() const { return RegisterType == 'x'; }
+  bool isSpaceOnly() const {
+    return RegisterType == 0 && RegisterSpace.hasValue();
```
Comment (ehsannas) [inline tools/clang/include/clang/AST/HlslTypes.h]: Actually, I wonder if `llvm::Optional` is necessary?

IIUC the absence of `space#` implies that space #0 should be used, so with `llvm::Optional`, we will end up using `RegisterSpace.getValueOr(0)` everywhere. Without `llvm::Optional`, we can initialize `RegisterSpace` to zero; only set its value when provided in the HLSL source; and use `RegisterSpace` as a uint everywhere.

And `isSpaceOnly` function can just `return RegisterType == 0;`

What do you think?

Comment (ehsannas) [reply tools/clang/include/clang/AST/HlslTypes.h]: (as a side: the fix for the SPIR-V backend is just a matter of adding `getValueOr` in some places I believe, and I wouldn't mind doing that if you think `llvm::Optional` makes more sense)

Comment (MrTrillian) [reply tools/clang/include/clang/AST/HlslTypes.h]: We have an `-auto-binding-space` parameter to choose the space for resources with no explicit assignment, which means that I must distinguish between `Buffer buf;` and `Buffer buf : register(space0);` because it would only apply to the former. I guess I could distinguish based on the presence of the RegisterAllocation UnusualAttribute, but since its space parameter is optional, the current data model maps better to the syntax.

Comment (ehsannas) [reply tools/clang/include/clang/AST/HlslTypes.h]: Ah! I see. Sorry I forgot about `-auto-binding-space`. Makes sense. I'll update the backend code. Thanks!

**Comment by MrTrillian** [comment]
Rebased on latest master.

**Comment by ehsannas** [comment]
LGTM. Thanks!

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp [hunk 2]
```diff
@@ -2853,6 +2845,13 @@ void CGMSHLSLRuntime::AddConstant(VarDecl *constDecl, HLCBuffer &CB) {
     case hlsl::UnusualAnnotation::UA_RegisterAssignment: {
       if (isGlobalCB) {
         RegisterAssignment *ra = cast<RegisterAssignment>(it);
+        if (ra->RegisterSpace.hasValue()) {
+          DiagnosticsEngine& Diags = CGM.getDiags();
+          unsigned DiagID = Diags.getCustomDiagID(
+            DiagnosticsEngine::Error,
+            "space is only allowed in a constant buffer.");
```
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGHLSLMS.cpp]: I don't think this is quite right.

First, the enclosing context is: we have a constant global variable declaration we wish to add to a constant buffer which may be inside explicit `cbuffer { ... }`, or outside - contributing to the `$Globals` default cbuffer.  `isGlobalCB` is true in the latter case.

I believe this case is triggered when supporting legacy c#, b#, and i# DX9-style register binding, which would define the constant index within the constant array for this variable.  The DX10+ equivalent is the offset within the $Globals constant buffer.  Since DX9 didn't have explicit `cbuffer {}` declarations, it only applies to the $Globals case.  Perhaps the code should at least warn when used inside `cbuffer {}`, since it's not clear that your register annotation will be ignored, but maybe there is an error or warning emitted earlier for this case.

One thing that is clear, space binding shouldn't be included here at all, since it's not a concept that applies within a constant buffer (making the error message incorrect).  In general, I don't think space should be set/allowed for any scenario handled inside this function (though see additional complications below).

There are additional complications for this case that I'm not sure are addressed, but not necessary to address as part of this change, such as:
- we really should only respect the c# bindings here, not b# or i#, since those addressed different constant banks in DX9.
- How does this code handle structures containing numeric types and resources?  Space would apply to resources in a struct in a constant buffer, but only for the resource binding portion.  This is a more widespread problem we need to fix though.
- How about multiple register bindings?

In summary, I think we should generally disallow space here, and figure out which unit tests go through these code paths to see what coverage we have, as well as add some likely missing unit tests.  If, for this PR, you want to get rid of the misleading diagnostic from the special case and add an issue for the follow-up work, I would understand.
